From 9826d1999ffdd75d52f5453837a787a126b02a3b Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Mon, 13 May 2013 18:54:20 +0900
Subject: [PATCH 6/9] ext/dfbvideosink: fix setting CbCr plane physical address
 of a framebuffer

CbCr plane physical addresses in several places are mistakenly set
to incorrect values.
---
 ext/directfb/dfbvideosink.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/ext/directfb/dfbvideosink.c b/ext/directfb/dfbvideosink.c
index 046183a..e7329d4 100644
--- a/ext/directfb/dfbvideosink.c
+++ b/ext/directfb/dfbvideosink.c
@@ -2324,9 +2324,9 @@ gst_dfbvideosink_shvio_stretchblit (GstDfbVideoSink * dfbvideosink,
 
     if (dfbvideosink->icbc[DST].handle) {
       gulong physc, pc_frac;
-      gulong chroma_byte_offset = 0;
 
-      physc = phys[DST] + chroma_byte_offset;
+      physc = phys[DST] + byte2pixel (dest_pitch, dst_format) *
+          (dfbvideosink->out_height - dst->y / 2);
       pc_frac = physc & 0x0000000f;
 
       meram_write_icb (dfbvideosink->meram, dfbvideosink->icbc[DST].handle,
@@ -2414,7 +2414,8 @@ gst_dfbvideosink_shvio_stretchblit (GstDfbVideoSink * dfbvideosink,
 #if defined(HAVE_SHVIO) && defined(HAVE_SHMERAM)
 static GstFlowReturn
 gst_dfbvideosink_shvio_fill (GstDfbVideoSink * dfbvideosink, guint8 * virt_addr,
-    GstVideoRectangle * rect, DFBSurfacePixelFormat format, gulong argb)
+    gint stride, GstVideoRectangle * rect, DFBSurfacePixelFormat format,
+    gulong argb)
 {
   struct ren_vid_surface viosurface;
   int ret;
@@ -2453,12 +2454,12 @@ gst_dfbvideosink_shvio_fill (GstDfbVideoSink * dfbvideosink, guint8 * virt_addr,
       dfbvideosink->icby[DST].handle, 0);
 
   if (dfbvideosink->icbc[DST].handle) {
-    gulong chroma_byte_offset = 0;
     meram_write_icb (dfbvideosink->meram, dfbvideosink->icbc[DST].handle,
         MExxBSIZE, (((rect->h / 2 - 1) & 0x1fff) << 16) | ((pixel2byte (rect->w,
                     format) - 1) & 0x7fff));
     meram_write_icb (dfbvideosink->meram, dfbvideosink->icbc[DST].handle,
-        MExxSSARA, phys + chroma_byte_offset);
+        MExxSSARA, phys + byte2pixel (stride, format) *
+        dfbvideosink->out_height);
     viosurface.pc = (void *)
         meram_get_icb_address (dfbvideosink->meram,
         dfbvideosink->icbc[DST].handle, 0);
@@ -2610,7 +2611,8 @@ gst_dfbvideosink_show_frame (GstBaseSink * bsink, GstBuffer * buf)
       dst.h = dfbvideosink->out_height;
       /* To clear MERAM, the destination surface is filled out with
          the background color through MERAM by this function. */
-      ret = gst_dfbvideosink_shvio_fill (dfbvideosink, data, &dst, dst_format,
+      ret = gst_dfbvideosink_shvio_fill (dfbvideosink, data, stride,
+          &dst, dst_format,
           (dfbvideosink->bgcolor.alpha << 24) |
           (dfbvideosink->bgcolor.red << 16) |
           (dfbvideosink->bgcolor.green << 8) | dfbvideosink->bgcolor.blue);
-- 
1.8.1.2

